### PR TITLE
fix some unsafe behavior in std.format

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

File: std/format.d [hunk 1]
```diff
@@ -6549,3 +6549,14 @@ unittest
...
     });
 }
+
+/*****************************
+ * The .ptr is unsafe because it could be dereferenced and the length of the array may be 0.
+ * Returns:
+ *      the difference between the starts of the arrays
+ */
+@trusted private pure nothrow @nogc
+    size_t arrayPtrDiff(const void[] array1, const void[] array2)
```
Comment (ZombineDev) [inline std/format.d]: Shouldn't this return `ptrdiff_t` or at least assert that `array1.ptr >= array2.ptr`?

Comment (burner) [inline std/format.d]: indeed

**Comment by burner** [comment]
What is made safe by this and where is the @safe unittest that shows that.

**Comment by ZombineDev** [comment]
@burner Walter is fixing various safety related bugs. In this case, the next version of dmd will disallow accessing `.ptr` in `@safe` code (which is currently allowed, IIRC). In order for the DMD PR to pass on the autotester, accessing `.ptr` should first be encapsulated throughout druntime ans phobos.

See also https://github.com/dlang/dmd/pull/5860

**Comment by burner** [comment]
@ZombineDev thanks. @WalterBright I would have liked this link to the dmd PR in the original PR description. Figuring out what the purpose of this PR is would have been easier this way.

**Comment by burner** [comment]
LGTM

File: std/format.d [hunk 2]
```diff
@@ -988,7 +988,7 @@ struct FormatSpec(Char)
                 const widthOrArgIndex = parse!uint(tmp);
                 enforceFmt(tmp.length,
                     text("Incorrect format specifier %", trailing[i .. $]));
-                i = tmp.ptr - trailing.ptr;
+                i = arrayPtrDiff(tmp, trailing);
```
Comment (ntrel) [inline std/format.d]: Alternatively, we could use a more general ptrValue safe function:
`i = tmp.ptrValue - trailing.ptrValue;`
See https://github.com/dlang/druntime/pull/1590#issuecomment-225712106

**Comment by schveiguy** [comment]
Auto-merge toggled on

File: std/format.d (hunk 2 above)
Comment (schveiguy) [inline std/format.d]: @ntrel see new PR I added to try and make a nice wrapper for this: https://github.com/dlang/druntime/issues/1592

**Comment by CyberShadow** [comment]
This pull request introduced a regression:
https://issues.dlang.org/show_bug.cgi?id=16676

**Pull request merged**